### PR TITLE
Add support for IoT Domain Configuration

### PR DESCRIPTION
**Pull request by llawford**
Summary: 
Added support for the IoT DomainConfiguration object, and the create/update/describe/list/delete operations.
Added tests covering each operation plus validation. 

Checklist:
- [x] Feature is added
- [x] The linter is happy
- [x] Tests are added      
- [x] All tests pass, existing and new

### CODE REVIEW

File: moto/iot/models.py [hunk 1]
```diff
@@ -484,6 +484,74 @@ def to_dict(self):
...
+        region_name,
+        domain_configuration_name,
+        domain_name,
+        server_certificate_arns,
+        domain_configuration_status,
+        service_type,
+        authorizer_config,
+        domain_type,
+    ):
+        if domain_configuration_status not in ["ENABLED", "DISABLED"]:
```
Comment (bblommers) [inline moto/iot/models.py:499]: This branch doesn't seem to be reachable, as we're always passing in a hardcoded `ENABLED` when creating an instance

File: moto/iot/models.py [hunk 2]
```diff
@@ -484,6 +484,74 @@ def to_dict(self):
...
+            raise InvalidRequestException(
+                " An error occurred (InvalidRequestException) when calling the DescribeDomainConfiguration "
+                "operation: Domain configuration status %s not recognized."
+                % domain_configuration_status
+            )
+        if domain_type not in ["ENDPOINT", "AWS_MANAGED", "CUSTOMER_MANAGED"]:
+            raise InvalidRequestException(
```
Comment (bblommers) [inline moto/iot/models.py:506]: Same here

File: moto/iot/models.py [hunk 3]
```diff
@@ -484,6 +484,74 @@ def to_dict(self):
...
+                " An error occurred (InvalidRequestException) when calling the DescribeDomainConfiguration "
+                "operation: Domain type %s not recognized." % domain_type
+            )
+        # Handle None service type (is it possible)?
+        if service_type and service_type not in ["DATA", "CREDENTIAL_PROVIDER", "JOBS"]:
```
Comment (bblommers) [inline moto/iot/models.py:511]: There should be a test for this branch

File: tests/test_iot/test_iot.py [hunk 4]
```diff
@@ -2346,3 +2346,115 @@ def test_update_thing_with_deprecated_type(self):
...
+
+    # Creation with only the name should succeed, creating same name fails
+    domain_config_1 = client.create_domain_configuration(
+        domainConfigurationName="testConfig1"
+    )
+    domain_config_1.should.have.key("domainConfigurationName").which.should.equal(
+        "testConfig1"
+    )
+    domain_config_1.should.have.key("domainConfigurationArn").which.should_not.be.none
+    with pytest.raises(client.exceptions.ResourceAlreadyExistsException):
```
Comment (bblommers) [inline tests/test_iot/test_iot.py:2363]: With negative tests, we always have to verify the format is correct:
```
with pytest.raises(..) as exc:
    ...
err = exc.value.response["Error"]
err["Code"].should.equal("..")
err["Message"].should.equal("..")
```

**Review by bblommers — CHANGES_REQUESTED** [review]
Hi @llawford, thanks for the PR!
Would you mind splitting  up the tests in multiple methods? Ideally one scenario per test, i.e. `test_create..`, `test_list..`, `test_create_with_invalid_params`, etc..

File: moto/iot/models.py (hunk 1 above)
Comment (llawford) [reply moto/iot/models.py:499]: You're right; domain configurations seem to always start as ENABLED and the domain type can't be set by the user, AWS sets it automatically based on the domain, which is why I hardcoded them. I'll remove these two checks, they can be added in the future if these parameters are allowed to be set somehow.

**Comment by llawford** [comment]
@bblommers Sorry for the multiple fix commits. Let me know if I'm okay to squash these commits at some point, not sure if the merge process does that automatically here.

**Comment by bblommers** [comment]
> @bblommers Sorry for the multiple fix commits. Let me know if I'm okay to squash these commits at some point, not sure if the merge process does that automatically here.

All good - we'll squash them anyway when merging!

Happy to merge once the CI passes - it looks like black is stil failing atm.

I see you already did some linting - keep in mind that different versions of black will give different results. We're using `black==19.10b0`, so only linting with that specific version guarantees that the build will pass.

**Review by bblommers — APPROVED** [review]
LGTM! Thanks again @llawford

**Pull request merged**